python/src/lib.rs

```diff
@@ -520,6 +525,67 @@ impl RawDeltaTable {
         Ok(())
     }
 
+    #[pyo3(signature = (starting_version, ending_version = None, starting_timestamp = None, ending_timestamp = None))]
+    pub fn read_cdf(
+        &mut self,
+        py: Python,
+        starting_version: i64,
+        ending_version: Option<i64>,
+        starting_timestamp: Option<String>,
+        ending_timestamp: Option<String>,
+    ) -> PyResult<PyArrowType<RecordBatch>> {
+        let ctx = SessionContext::new();
+        let mut cdf_read = CdfLoadBuilder::new(
+            self._table.log_store(),
+            self._table.snapshot().map_err(PythonError::from)?.clone(),
+        )
+        .with_starting_version(starting_version);
+
+        if let Some(ev) = ending_version {
+            cdf_read = cdf_read.with_ending_version(ev);
+        }
+        if let Some(st) = starting_timestamp {
+            let starting_ts: DateTime<Utc> = DateTime::<Utc>::from_str(&st)
+                .map_err(|pe| PyValueError::new_err(pe.to_string()))?
+                .to_utc();
+            cdf_read = cdf_read.with_starting_timestamp(starting_ts);
+        }
+        if let Some(et) = ending_timestamp {
+            let ending_ts = DateTime::<Utc>::from_str(&et)
+                .map_err(|pe| PyValueError::new_err(pe.to_string()))?
+                .to_utc();
+            cdf_read = cdf_read.with_starting_timestamp(ending_ts);
+        }
+
+        cdf_read = cdf_read.with_session_ctx(ctx.clone());
+
+        let plan = rt().block_on(cdf_read.build()).map_err(PythonError::from)?;
+
+        py.allow_threads(|| {
+            let mut tasks = vec![];
+            for p in 0..plan.output_partitioning().partition_count() {
+                let inner_plan = plan.clone();
+                let partition_batch = inner_plan.execute(p, ctx.task_ctx()).unwrap();
+                let handle = rt().spawn(collect_sendable_stream(partition_batch));
+                tasks.push(handle);
+            }
+
+            // This is unfortunate.
+            let batches: Vec<_> = rt()
+                .block_on(join_all(tasks))
+                .into_iter()
+                .flatten()
+                .collect::<Result<Vec<Vec<_>>, _>>()
+                .unwrap()
+                .into_iter()
+                .flatten()
+                .collect();
+
+            let final_batch = concat_batches(&batches[0].schema(), &batches).unwrap();
+            Ok(PyArrowType(final_batch))
```

Comment:
I think we can also convert it as a reader again

Reply:
I'm sorry, what do you mean?

Reply:
impl IntoPyArrow for ArrowArrayStreamReader
Convert a ArrowArrayStreamReader into a pyarrow.RecordBatchReader.

https://docs.rs/arrow/latest/arrow/ffi_stream/struct.ArrowArrayStreamReader.html

I mean the above. You can take an arrowArrayStreamReader and move it into python as pyarrow.RecordBatchReader

Reply:
Create a RecordBatchReader first from the `batches` then the stream

```
let stream = FFI_ArrowArrayStream::new(reader)
Ok(PyArrowType(stream))
```

With this as the return type: PyResult<PyArrowType<ArrowArrayStreamReader>>